design/two-stage-backup-api-proposal.md

```diff
@@ -0,0 +1,149 @@
+           |
+           |
+           | < 2 >
+           |
+           v
++----------+---------+
+|                    |
+|    Post-Hook       |
+|                    |
++--------------------+
+
+
+
+
+```
+
+In the above flow, `CreateSnapshot` method creates snapshot and starts upload too. Hence, split this into two APIs: `CreateSnapshot` and `UploadSnapshot`.
+
+Proposed design: `Pre-hook` -> `CreateSnapshot` -> `Post-hook` -> `UploadSnapshot` (Background)
+
+```acsii
+
++--------------------+
+|                    |
+|     Pre-Hook       |
+|                    |
++---------+----------+
+          |
+          |
+          | < 1 >
+          |
+          v
++---------+----------+
+|                    |
+|   CreateSnapshot   |
+|                    |
++----------+---------+
+           |
+           |
+           | < 2 >
+           |
+           v
++----------+---------+
+|                    |
+|    Post-Hook       |
+|                    |
++----------+---------+
+           |
+           | < 3 >
+           |
+           v
+ +---------+---------+
+ |                   |
+ |  UploadSnapshot   |
+ |   (Background)    |
+ +-------------------+
+
+
+```
+
+## Detailed Design
+
+Velero has a pretty neat feature to maintain the application consistent backups/snapshot called `Pre-hooks` and `Post-Hooks`. However, there's a small change in Velero which needs to be addressed in order to achieve minimum downtime for an application or more consistent application backups. Existing backup flow for a typical stateful application looks like: `Pod` -> `Pre-Hook` --> `PVC` -> `PV` -> `Post-Hook`. This is a mechanism which is followed by Velero block store plugins. In most of the plugins, `CreateSnapshot` method of `VolumeSnapshotter` interface implements logic to upload the snapshot and this is a blocking call for `Post-Hook`. This will introduce more time and will keep application frozen for time equals to snapshot upload time.
+
+
+To solve the above problem, we can split existing backup flow and introduce one more method called `UploadSnapshot` to `VolumeSnapshotter` interface, which will execute once `Post-Hook` is executed and will keep running in background. This will reduce the time between `Pre-Hook` and `Post-Hook` significantly as well as it will not impact application consistency or downtime.
+
+### Implementation
+
+This change should be implemented in both Velero as well as Velero plugins. However, Velero plugins can continue to rely on current implementation. Only way for Velero plugins to get benifitted from this approach is to implement `UploadSnapshot` method.
+
+Volumesnapshotter is designed for snapshot, this can be either cloud snapshot or local snapshot. As of now, most plugin use volumesnapshotlocations to decide the snapshot type.
+We can add new config parameter in velero command line to define snapshot type, through this, velero will decide whether to invoke `UploadSnapshot` or not.
+
+**Implementation plan for Velero:**
```

Comment:
> we probably want a timeout value for the UploadSnapshot calls. Do you think this should be configurable on the VolumeSnapshotLocation?

This is not clear to me. Can you be specific about, why do we want timeout? Is it a timeout or number of retries for `UploadSnapshot` call?

> what should happen if 1+ UploadSnapshot calls fail?

We should mark backup as `Failed` since, snapshot succeeded but upload failed. If we are not able to reuse (upload again) the snapshot then it doesn't make much sense to me to mark entire backup as `PartiallyFailed`. 
If we are marking backup as `Failed` in the above case then we must have intelligence to do clean up too. Otherwise, snapshot will unnecessarily be present on the node. 

> Do we need some way to delete the local snapshot for cleanup? 

Yes! For local snapshot we should provide a way to do clean-up. Clean up can be triggered if `UploadSnapshot` returns an error. Clean up can be a go routine too (Part of same waitGroup). 

(On the similar line, I was thinking to use channel for passing signal from `UploadSnapshot` to  trigger a `DeleteSnapshot` call but that would require to change the signature of `UploadSnapshot` which wouldn't be good since there are third party plugins which will consume  `VolumeSnapshotter`)

Reply:
@dsu-igeek You have put very good points here.

@skriss  I think, we should provide some configuration for clean up too or skip the clean up part for now and make sure that backup is marked as `PartiallyFailed` if upload fails in between. I need your help on deciding this.

@nrb 
> If you delete the local snapshot, how does restore work? Do we need a download method that pulls the snapshot out of the long term storage to use?

I really haven't given a thought to this side of the problem. However, your suggestion sounds right to me. I think, we should first address the backup part. Restore will come into picture when we are doing clean up of existing snapshots. 

@skriss I think, we should consider clean up in the second phase. I see that, snapshot can be used even if snapshot is uploaded to remote storage. This will require modifications on restore side.

Reply:
>I think, we should first address the backup part.

That's reasonable. I think we can certainly iterate on the backup design first.

My personal preference would be to not release the API until we're fairly confident in both backup and restore, to minimize the number of new releases plugins would have to make to conform.